Stock.playground/Contents.swift

```diff
@@ -3,116 +3,147 @@ import Foundation
     }
 
-    // NSOs
-    init (shares: UInt, grantDate: GrantDate, cliffMonths: UInt, endMonths: UInt, strikePrice: Double)
+    // Bonus grants: vest instantly
+    init(shares: UInt, grantDate: Date, strikePrice: Double = 0)
     {
-        self.strikePrice = strikePrice
         self.shares = shares
+        self.strikePrice = strikePrice
         self.grantDate = grantDate
-        self.cliffMonths = cliffMonths
-        self.endMonths = endMonths
+        cliffMonths = 0
+        endMonths = 0
     }
 
-    func valueAfter(months: Int, atPrice: Double) -> Double
+    func valueAt(_ price: Double, monthsElapsed months: UInt) -> Double
     {
-        guard months >= Int(self.cliffMonths) else { return 0 }
-        guard months <= Int(self.endMonths) else { return self.valueAfter(Int(self.endMonths), atPrice: atPrice) }
-        guard atPrice > self.strikePrice else { return 0 }
+        guard months >= cliffMonths else { return 0 }
+        guard months <= endMonths else { return valueAt(price, sharesAccrued: Double(shares)) }
+        guard price > strikePrice else { return 0 }
 
-        let shareValue = atPrice - self.strikePrice
-        let sharesAccrued = floor(Double(months) / Double(self.endMonths) * Double(self.shares))
-        return shareValue * sharesAccrued
+        let percentageAccrued = Double(months) / Double(endMonths)
+        let sharesAccrued = floor(percentageAccrued * Double(shares))
+        return valueAt(price, sharesAccrued: sharesAccrued)
     }
 
-    func monthsSince(date: GrantDate) -> Int
+    func valueAt(_ price: Double, date: Date) -> Double
     {
-        let futureDate = date.foundationDate
-        let grantDate = self.grantDate.foundationDate
-
-        let dateComponents = NSCalendar.currentCalendar().components(NSCalendarUnit.Month, fromDate: grantDate, toDate: futureDate, options: NSCalendarOptions(rawValue: 0))
-        return dateComponents.month
+        let monthsElapsed = date.monthsSince(grantDate)
+        if (monthsElapsed < 0) { return 0 }
+        return valueAt(price, monthsElapsed: UInt(monthsElapsed))
     }
 
-    func valueBy(date: GrantDate, atPrice: Double) -> Double
+    private func valueAt(_ price: Double, sharesAccrued: Double) -> Double
     {
-        return self.valueAfter(self.monthsSince(date), atPrice: atPrice)
+        let shareValue = price - strikePrice
+        let value = max(0, shareValue)
+        let shareCount = max(0, min(Double(shares), sharesAccrued))
+        return value * shareCount
     }
+}
+
 
-    func valuesBy(dates: [GrantDate], atPrice: Double) -> [Double]
+extension Date
+{
+    static func grantDate(month: Int, day: Int, year: Int) -> Date
     {
-        return dates.map({ self.valueBy($0, atPrice: atPrice) })
+        let signing = NSDateComponents.init()
+        signing.calendar = Calendar.grantDateCalendar()
+        signing.month = month
+        signing.day = day
+        signing.year = year
+        return signing.date!
```

Comment:
we can do better! No `!`. Probably best for this method to return an optional, since the inputs aren't pre-validated.